From:  timothy shimmin <tes@sgi.com> (from "Lachlan")
Subject: Fix propagation of EOPNOTSUPP when barriers are not supported.
Patch-mainline: 2.6.28
References: bnc#438608

If a device reports EOPNOTSUP because a barrier was requested but
not supported, that error currently gets transformed into EIO.
This fix keeps the correct error.

Acked-by: Neil Brown <neilb@suse.de>
Signed-off-by: Neil Brown <neilb@suse.de>

---
 fs/xfs/linux-2.6/xfs_buf.c |    3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

--- linux-2.6.27.orig/fs/xfs/linux-2.6/xfs_buf.c
+++ linux-2.6.27/fs/xfs/linux-2.6/xfs_buf.c
@@ -1114,8 +1114,7 @@ xfs_buf_bio_end_io(
 	unsigned int		blocksize = bp->b_target->bt_bsize;
 	struct bio_vec		*bvec = bio->bi_io_vec + bio->bi_vcnt - 1;
 
-	if (!test_bit(BIO_UPTODATE, &bio->bi_flags))
-		bp->b_error = EIO;
+	xfs_buf_ioerror(bp, -error);
 
 	do {
 		struct page	*page = bvec->bv_page;
